Rails views

Yuliya-Kirich opened the pull request:
1. implemented all standard actions and views for resource question. 
2. added footer, which displays the current year, author's name, project name. 
3. added a helper for the title of the page for creating and editing a question. 
added date helper in footer. 
added helper with links to author and project.

Yuliya-Kirich committed:
1. implemented all standard actions and views for resource question. 2. added footer, which displays the current year, author's name, project name. 3.added a helper for the title of the page for creating and editing a question. Added date helper in footer.
Added helper with links to author and project.

Yuliya-Kirich committed:
Revert "1. implemented all standard actions and views for resource question. 2. added footer, which displays the current year, author's name, project name. 3.added a helper for the title of the page for creating and editing a question. Added date helper in footer."

This reverts commit 472e4cfe9910c21c54942f11767cb6986706c7f9.

Yuliya-Kirich committed:
Revert "Revert "1. implemented all standard actions and views for resource question. 2. added footer, which displays the current year, author's name, project name. 3.added a helper for the title of the page for creating and editing a question. Added date helper in footer.""

This reverts commit 9a00173329fb9b7f4f52e7d7a0c528b56ab1dff2.

VidgarVii commented on Gemfile:
```diff
@@ -55,4 +55,5 @@ gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
-gem 'rails-i18n'
\ No newline at end of file
+gem 'rails-i18n'
+#gem 'pagy'
```
в конце файла надо отставлять 1 пустую строчку. Это указатель на конец файла. Сейчас система думает, что происходит обрыв файла. Даже гитхаб ругается. No new lina at end of file

VidgarVii replied:
этим гемом не будем пользоваться. Лучше его удалить полностью, а не прятать за комментом

VidgarVii commented on app/controllers/questions_controller.rb:
```diff
@@ -1,35 +1,49 @@
 
   def index
-    @question = @test.questions
+    @questions = @test.questions
```
важно:
ТЗ: "Список вопросов теста должен выводиться **на той же странице, что и сам тест**"
другими словами нужна страница для вывода теста
что приводит к следующим выводам:
- экшен `index` в классе `QuestionsController` лишний
- вызов коллбека для `index` не нужен
- список вопросов нужно выводить на странице теста
- путь `test_questions_path`, который ты используешь в экшенах, тоже лишний (нужно отменить его создание в `routes`)
- нужен контроллер для ресурса `:tests` (его написание было поэтапно продемонстрировано в скринкасте)

VidgarVii commented on app/controllers/questions_controller.rb:
```diff
@@ -1,35 +1,49 @@
   end
 
   def create
     @question=@test.questions.new(question_params)
```
не забываем про пробельчики
`@question = @test.questions.new(question_params)`

VidgarVii commented on app/controllers/questions_controller.rb:
```diff
@@ -1,35 +1,49 @@
     @question=@test.questions.new(question_params)
     if @question.save
-      render action: :show
+      redirect_to action: :show, id: @question.id
```
можно написать короче
`redirect_to @question`

VidgarVii commented on app/controllers/questions_controller.rb:
```diff
@@ -1,35 +1,49 @@
   end
 
   def show
+    #@test= @question.test
```
что это?)

VidgarVii commented on app/controllers/questions_controller.rb:
```diff
@@ -1,35 +1,49 @@
      else
-      render json: question.errors
+      render json: @question.errors
      end
```
обычно ресурсы должны удаляться без проблем. Если это не так, то это проблема кода, а не юзера

VidgarVii commented on app/controllers/tests_controller.rb:
```diff
@@ -1,43 +1,69 @@
 
   def index
-    @test=Test.all
+    @test=Test.order(params[:sort])
```
пробелы. Надо следить за чистотой кода. 
Используй rubocop для автоматизации процесса.
Еще есть инструменты, которые будут запускать линтер автоматически

VidgarVii commented on app/controllers/tests_controller.rb:
```diff
@@ -1,43 +1,69 @@
+    # @test = Test.new(user_id: 1)
+    @categories = Category.group(:number)
+    @test= Test.new(test_params)
+    @test.user_id=1
```
стоит привыкнуть работать с объектами, а не с их атрибутами.
И если нет такого юзера с ИД = 1?
```ruby
author = User.first
@test = author.authored_tests.build(test_params)
```

VidgarVii commented on app/controllers/tests_controller.rb:
```diff
@@ -1,43 +1,69 @@
+    else
+      render :new
+    end
+    #render action: :show, id: @test.id
```
такие комментарии не нужны в проекте. Надо удалять за собой

VidgarVii commented on app/helpers/application_helper.rb:
```diff
@@ -1,2 +1,8 @@
+    content_tag(:p,  "2021 - #{Time.now.strftime('%Y')}")
+  end
+  def github_url(author, repo)
+    "<p style='text-align: center'><a href='#{author}'>Информация об авторе: Кириченко Юлия. </a><a href='#{repo}'>проект Тест Гуру</a></p>".html_safe
```
есть удобный хелпер link_to для ссылок
https://apidock.com/rails/ActionView/Helpers/UrlHelper/link_to

VidgarVii commented on app/helpers/tests_helper.rb:
```diff
@@ -1,2 +1,7 @@
 module TestsHelper
+  TEST_LEVELS={0=>"Легкий", 1=>"Средне-легкий", 3=>"Средний", 4=>"Средне-сложный", 5=>"Сложный", 6=>"Очень сложный"}.freeze
+  def test_level(test)
```
м/у методом и константой нужен отступ

VidgarVii commented on app/helpers/tests_helper.rb:
```diff
@@ -1,2 +1,7 @@
 module TestsHelper
+  TEST_LEVELS={0=>"Легкий", 1=>"Средне-легкий", 3=>"Средний", 4=>"Средне-сложный", 5=>"Сложный", 6=>"Очень сложный"}.freeze
```
массив это тот же хеш у которого ключи в виде цифр
`TEST_LEVELS = %w[Легкий Средне-легкий Средний ....].freeze`
так смотрится на много лучше
и метод test_level переписывать не придётся

VidgarVii commented on app/helpers/tests_helper.rb:
```diff
@@ -1,2 +1,7 @@
+  TEST_LEVELS={0=>"Легкий", 1=>"Средне-легкий", 3=>"Средний", 4=>"Средне-сложный", 5=>"Сложный", 6=>"Очень сложный"}.freeze
+  def test_level(test)
+    TEST_LEVELS[test.level] || "В процессе создания"
+  end
```
можно еще написать так.
`TEST_LEVELS.fetch(test.level, 'В процессе создания')`
выглядит более декларативно

VidgarVii commented on app/views/layouts/application.html.erb:
```diff
@@ -7,9 +7,18 @@
   </body>
-</html>
+
+<footer>
```
чтобы layouts остался чистым футер лучше положить в партиал

VidgarVii commented on app/views/questions/_form.html.erb:
```diff
@@ -0,0 +1,28 @@
+<div class="rails-default-about-page">
+  <h1><%=question_header%></h1>
+  <%if action_name == 'edit' %>
```
это лишняя проверка
_Для того чтобы метод form_with смог сформировать верный URL используйте вызов form_with model: [@parent, @child]_
просто пишешь `form_with model: [@test, @question]`
для эдит переменная `@test` будет пустая (не надо ее в контроллере искать, т.к. при использовании опции "шеллоу тру" в роутах для эдит `@test` не нужен)
соответсвенно в `@test` будет "нил" и рельсы построят правильный путь, потому что придет только `@question`
для нью придут оба объекта ("тест" и "квесшен") и рельсы также построят правильный путь

VidgarVii commented on app/views/shared/_question.html.erb:
```diff
@@ -0,0 +1,21 @@
+  <th>Дата обновления</th>
+  <th> </th>
+</tr>
+<%n=1 %>
```
это зачем?

VidgarVii commented on app/views/tests/index.html.erb:
```diff
@@ -1,40 +1,18 @@
-    </tr>
-      </tbody>
-    <%end%>
+  <%=render @test%>
```
`<%= render @test %>`

Yuliya-Kirich committed:
fix 1

VidgarVii commented on app/views/questions/index.html.erb:
```diff
@@ -1,16 +0,0 @@
-<div class="rails-default-about-page">
```
тут можно весь файл удалить, а не только код

VidgarVii commented on app/views/tests/new.html.erb:
```diff
@@ -1,4 +1,4 @@
   <h1>Создать тест</h1>
-  <%=render 'form' %>
+  <%= render 'form' %>
 </div>
```
в конце файла надо отставлять 1 пустую строчку. Это указатель на конец файла. Сейчас система думает, что происходит обрыв файла. Даже гитхаб ругается. No new lina at end of file

VidgarVii commented on app/views/tests/show.html.erb:
```diff
@@ -1,33 +1,29 @@
   <%if @test.questions.present?%>
   <table  class="table">
-    <%= render 'shared/question', resource:[@test, @question]%>
+    <%= render 'shared/question', resource:[@test, @question] %>
```
для рессурса вопросов есть уже папка questions
Там по умолчанию и должен лежить паршал для показа вопроса
достаточно было сделать так
`<%= render @test.questions %>`

Yuliya-Kirich committed:
fix routes

VidgarVii commented on app/helpers/questions_helper.rb:
```diff
@@ -2,10 +2,10 @@ module QuestionsHelper
+          @test.title
       end} "), class: "strong")
   end
 end
```
https://github.com/Yuliya-Kirich/thinknetika_base/pull/12#discussion_r1001581076
Многие компиляторы тут выдадут ошибку. Руби частично компилируемый язык. И такой разрыв легко проглотит. Но в реальных проектах это как красная тряпка для быка. Так что тренируем дисциплину и больше такого не совершаем! В руби за чистоту кода!

VidgarVii commented on app/helpers/questions_helper.rb:
```diff
@@ -2,10 +2,10 @@ module QuestionsHelper
   def question_header
```
как этот код попал в мастер? хелпер пишется только на этом задании

VidgarVii commented on app/helpers/questions_helper.rb:
```diff
@@ -2,10 +2,10 @@ module QuestionsHelper
+          @question.test.title
+        else
+          @test.title
       end} "), class: "strong")
```
не стоит полагаться на глобальное состояние (инстанс-переменные) в хелпере
лучше передавать вопрос параметром, иначе потом не разберешь откуда эти данные тут взялись
так же легко забыть, что здесь используется эта инстанс-переменная, переименовать ее при рефакторинге и сломать код

Yuliya-Kirich replied:
https://github.com/Yuliya-Kirich/thinknetika_base/pull/12/commits/9678029e57072aa54c722eb24a0609f609395f97

VidgarVii commented on app/helpers/application_helper.rb:
```diff
@@ -1,2 +1,8 @@
 module ApplicationHelper
-end
+  def footer_year
+    content_tag(:p,  "2021 - #{Time.now.strftime('%Y')}")
```
в рельсах рекомендуется использовать `Time.current` вместо `Time.now` (`Time.new`) и `Date.current` вместо `Date.today`
`.current` возвращает текущий момент времени с учетом настройки часового пояса в приложении

Yuliya-Kirich replied:
https://github.com/Yuliya-Kirich/thinknetika_base/pull/12/commits/4146389e622405ab425418e3a30046dcacdf4ccf

VidgarVii commented on app/views/questions/show.html.erb:
```diff
@@ -1,38 +1,19 @@
-  <h3></h3>
-
+  <h1>Вопрос для теста "<%=@question.test.title%>"</h1>
   <form action = "/questions/<%=@question.id%>" method="get">
```
не уверен что тут нужна форма. Форму показываем только при обновлении или создании вопроса

Yuliya-Kirich replied:
https://github.com/Yuliya-Kirich/thinknetika_base/pull/12/commits/4146389e622405ab425418e3a30046dcacdf4ccf

Yuliya-Kirich committed:
fix by rubocop and fix helper

Yuliya-Kirich committed:
fix helper

Yuliya-Kirich committed:
fix time in application_helper. delete form in show view by the questions.

VidgarVii commented on app/controllers/tests_controller.rb:
```diff
@@ -1,47 +1,70 @@
+    @test.questions.clear
+    @test.users.clear
+    @test = @test.destroy
    if @test.destroyed?
```
если рессур не может рекурсивно удалиться, то это баг. Весь код выше и проверки тут лишние
@test.destroy должен работать не зависимо от наличии дочерних объектов
соответственно экшен должен выглядеть примерно так

```ruby
@test.destroy

redirect_to tests_path
```

VidgarVii commented on app/views/tests/show.html.erb:
```diff
@@ -1,36 +1,36 @@
+      <th>Дата обновления</th>
+      <th> </th>
+    </tr>
+    <%= render @test.questions.first %>
```
по ТЗ надо показать все вопросы теста

VidgarVii commented on app/views/tests/show.html.erb:
```diff
@@ -1,36 +1,36 @@
-        <p><button class="button" onClick="window.open('/tests/<%=@test.id%>/questions/new')">Создать вопрос для теста</button></p>
-      </div>
+  <h1>Тест №<%= @test.id %></h1>
+  <%= form_with model: @test, local: true do |form| %>
```
для обновления теста, есть страница edit
Так же есть отдельный паршал для формы.
Хотя такой подход возможен, но мне кажется страница перегружена

VidgarVii commented on config/locales/models/rb.yml:
```diff
@@ -0,0 +1,16 @@
+        title: "Наименование теста"
+        level: "Уровень теста"
+        category_id: "Категория"
+        user_id: "Пользователь"
```
не исправлено

VidgarVii commented on config/locales/ru.yml:
```diff
@@ -1,4 +1,5 @@
+  questions_new: "Создать новый"
   clients:
     index:
       title: Список клиентов
```
не исправлено

Yuliya-Kirich committed:
fix: 1. model and controllers for destroy action; 2. locales for translate errors; 3. view question

Yuliya-Kirich committed:
fix 2

Yuliya-Kirich commented:
1. implemented all standard actions and views for resource question. 
2. added footer, which displays the current year, author's name, project name. 
3. added the helper for the title of the page for creating and editing a question.
4. added date helper in footer.

Yuliya-Kirich committed:
:q

Merge branch 'master' into rails_views
^E
